aerospike-core/src/net/connection_pool.rs

```diff
@@ -200,14 +217,14 @@ pub struct PooledConnection {
 impl PooledConnection {
     pub fn invalidate(mut self) {
         let conn = self.conn.take().unwrap();
-        self.queue.drop_conn(conn);
+        block_on(self.queue.drop_conn(conn));
     }
 }
 
 impl Drop for PooledConnection {
     fn drop(&mut self) {
         if let Some(conn) = self.conn.take() {
-            self.queue.put_back(conn);
+            block_on(self.queue.put_back(conn));
```

Comment:
Using futures::executor::block_on underneath an async function can cause a deadlock.

We had a production outage this week because of this line. I'm about to make a special patch to this branch to address it.

Reply:
Sorry for this inconvenience, i didnt know about that.
You should now be able to directly push to my branch in case its needed, since I currently dont have the time to work on this branch myself.

Reply:
Actually, I'll run my fix past you and @khaf to see if everyone's onboard with it first.

I changed internals to a _synchronous_ mutex, to allow it to be used more easily in the drop function.
Synchronous mutexes are safe in async code (and actually outperform async mutexes) with the proviso that they are never held across .await points... otherwise they deadlock.

So the put_back function now looks like this:

```
    pub fn put_back(&self, mut conn: Connection) {
        let mut internals = self.0.internals.lock().unwrap();
        if internals.num_conns < self.0.capacity {
            internals.connections.push_back(IdleConnection(conn));
        } else {
            aerospike_rt::spawn(async move { conn.close().await });
            internals.num_conns -= 1;
        }
    }
```


I also had to restructure `fn get(&self)` to do the synchronous code first and the asynchonous code later, as so:
```
    pub async fn get(&self) -> Result<PooledConnection> {
        let mut connections_to_free = Vec::new();
        let mut conn: Option<Connection> = None;
        {
            let mut internals = self.0.internals.lock().unwrap();
            while let Some(IdleConnection(this_conn)) = internals.connections.pop_front() {
                if this_conn.is_idle() {
                    connections_to_free.push(this_conn);
                    internals.num_conns -= 1;
                } else {
                    conn = Some(this_conn);
                    break;
                }
            }

            if conn.is_none() {
                if internals.num_conns >= self.0.capacity {
                    bail!(ErrorKind::NoMoreConnections);
                }
    
                internals.num_conns += 1;
            }
        }

        for mut connection in connections_to_free {
            connection.close().await;
        }

        if conn.is_none() {
            let new_conn = aerospike_rt::timeout(
                Duration::from_secs(5),
                Connection::new(&self.0.host.address(), &self.0.policy),
            )
            .await;

            let Ok(Ok(new_conn)) = new_conn else {
                self.0.internals.lock().unwrap().num_conns -= 1;
                bail!(ErrorKind::Connection(
                    "Could not open network connection".to_string()
                ));
            };

            conn = Some(new_conn);
        }

        Ok(PooledConnection {
            queue: self.clone(),
            conn,
        })
```

Reply:
Oh by the way. I mentioned the outage, but I didn't mention how great it is to have this patch at all. I'm not sure if we'd have a production environment without it.